### ISSUE TITLE
How to resolve Auto IAM AuthN Configuration Error?

### ISSUE THREAD

**Issue by BartStikkers** [GoogleCloudPlatform/cloud-sql-proxy#1679]
### Question

A few of my users are on Macs, and they can't seem to connect to our Cloud SQL Postgresql instance using the Auth Proxy, even though it's working fine for me on Windows (and for at least one other user on an M1 Mac). We're following the instructions on [this page](https://cloud.google.com/sql/docs/postgres/connect-auth-proxy), but connecting in a db manager using IAM credentials fails (we've tried several). The proxy indicates that it's running, but when we try to open the connection it shows that it's opening and then immediately closing connections on a few random ports, and the db manager (dbeaver CE, pgAdmin 4) says "FATAL: Cloud SQL IAM user authentication failed for user <and then the username in double quotes>". We've got Cloud CLI up, running, and authenticated; I've triple checked that user permissions were correct both in IAM and the SQL instance. 

**What are we missing?**

I'll note by the way that this version of the proxy does not have the command line option "-enable_iam_login". Instead, it seems to require to use "--gcloud-auth".

**Comment by enocom** [comment]
Can you show me how you're invoking the proxy on a machine that works and how you're doing it on a machine that doesn't?

If you're logging in using Auto IAM AuthN, you'll need to:

1. Make sure the credentials used by the Proxy match those of the IAM user in the database
2. Use either Application Default Credentials for a user account, or a credentials file for a service account
3. Launch the proxy like so:
``` shell
# with a credentials file
cloud-sql-proxy --auto-iam-authn <INSTANCE_CONNECTION_NAME> --credentials-file /path/to/key.json

# with a user's ADC
cloud-sql-proxy --auto-iam-authn <INSTANCE_CONNECTION_NAME>
```

**Comment by BartStikkers** [comment]
Thanks for getting back to me so quickly. Unfortunately, I didn't record anything and I don't have a Mac around to test it on. The people who do have their own stuff to be working on so I can't keep pestering them. I'll just recount the steps to the best of my memory, I hope that suffices.

I should note though, that I only found out about the version bump after I posted the question. It would be nice if that were mentioned in the [google cloud documentation](https://cloud.google.com/sql/docs/postgres/connect-auth-proxy), that would've saved us some confusion. The person who had it working on a Mac M1 was using a V1.xx version that he downloaded just before the page was changed.
### What we did

- installed the Cloud SDK / CLI
- `gcloud init`
- `gcloud auth login`
- downloaded the Auth Proxy from [here](https://cloud.google.com/sql/docs/postgres/connect-auth-proxy) and followed the steps appropriate for the relevant OS
- reboot
- what works:
  - all OSes: V1.xx: `cloud_sql_proxy -enable_iam_login -instances=<INSTANCE_CONNECTION_NAME>=tcp:<PORT>`
  - only on Windows: V2.xx: `cloud_sql_proxy --port <PORT> <INSTANCE_CONNECTION_NAME> --auto-iam-authn`
- what _does_ work on an Intel Mac: 
  - V2.xx: `cloud_sql_proxy --port <PORT> <INSTANCE_CONNECTION_NAME> --gcloud-auth`
  - and then logging into the database with a SQL account rather than the IAM account.
- I should note that we haven't tested V2 on an M1 Mac yet.

The workaround that I've landed on for now is to have people grab the latest V1.xx release from the Github releases page. It requires a bit more handholding from me, but it doesn't affect so many users that it's an immediate problem.

**Comment by jackwotherspoon** [comment]
Hi @BartStikkers, thanks for the additional details! 😄 

I believe the issue you are facing is a result of a bug in our documentation on the [Connect using the Cloud SQL Auth Proxy](https://cloud.google.com/sql/docs/postgres/connect-auth-proxy) that has to do with setting the credentials used by the proxy.

**TLDR**

You should use `gcloud auth application-default login` to set credentials to be used by the proxy (and user to be used for Auto IAM AuthN) instead of `gcloud auth login`. 
- `gcloud auth login` is to set credentials for a user to be able to perform gcloud commands
- `gcloud auth application-default login` is to set credentials for a user to be used by application developers using client libraries (such as the Cloud SQL Proxy).

Step in the docs where bug is:  
> [4. Credentials from an authenticated gcloud CLI client](https://cloud.google.com/sql/docs/postgres/connect-auth-proxy#credentials-from-an-authenticated-gcloud-cli-client.)

This step prompts users to run `gcloud auth login` to set credentials to be used by the Auth Proxy. However, this command only sets the user credentials to be used by `gcloud` commands and does not set credentials to be used by applications (or client libraries such as the proxy).

The Cloud SQL Proxy sources credentials from the environment's [Application Default Credentials](https://cloud.google.com/docs/authentication#adc) and the proper command to run instead of `gcloud auth login` is:

```
gcloud auth application-default login
```

Let me know if running the new command resolves your problems! 😄 I have filed a bug against our docs to have the credentials steps updated to showcase the proper command.

Have a great day!

**Comment by BartStikkers** [comment]
Okay, thanks, that makes sense. It's the end of the day here now though, so I'll revisit this with my team on Monday. I'm only left wondering why the new version works on my Windows machine without having run `gcloud auth application-default login`. Quite possibly, I've set ADC for some other reason at some point, or do they get set behind the scenes when one runs the old version?

**Comment by jackwotherspoon** [comment]
@BartStikkers You are most likely correct in that you probably set ADC credentials in the past and your environment is still picking them up to be used by the proxy.

You can test this theory by running `gcloud auth application-default revoke` on your machine and then try invoking the Cloud SQL proxy and see if it begins to fail.

If it does then you must of had ADC credentials set in your environment previously.

**Comment by jackwotherspoon** [comment]
@BartStikkers The [step in the docs](https://cloud.google.com/sql/docs/postgres/connect-auth-proxy#credentials-from-an-authenticated-gcloud-cli-client.) has been fixed to showcase the proper `gcloud auth application-default login` command. Thanks for bringing this to our attention and helping make the Cloud SQL Proxy experience a better one 😄 

I am going to close this as I believe this command was the root cause of your issues. If you are still facing issues feel free to re-open this and I will take another look.

Have a wonderful day!
